### PR TITLE
keyfood

### LINKED ISSUE #158: Key Food
http://keyfood.mywebgrocer.com/StoreLocator.aspx?f=cir&_ga=1.172156139.912115260.1485884954&cc=1&s=398059791&g=7a1c3fa4-720b-4363-a6cf-38b88b9596db&uc=581EF7#key-food

### PR DESCRIPTION
**Pull request by bealbrown**
Fixes #158

Testing basic functionality with Travis CI and spider bot. Will have to format hours and adjust formatting in properties dict.

### CODE REVIEW

**Comment by bealbrown** [comment]
Looks like the scraperbot had an issue with Twisted: 

downloader/exception_count': 1,
 'downloader/exception_type_count/twisted.internet.error.TCPTimedOutError': 1,

My local copy appears to be scraping correctly. Shall I close the PR and reopen it to try again?

**Comment by bealbrown** [comment]
Looks like the scraper is running correctly now, but the stores aren't displaying properly on map. Probably a formatting issue with address, or properties dict doesn't have everything it needs to do geocoding

File: locations/spiders/keyfood.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+	def parse(self, response):
+
+		# works ok for now, but should probably just populate with actual contents of dropdown, in case they add new states.
+		states = ['CT','NJ','NY','PA']
+
+		for state in states:
+
+			yield scrapy.FormRequest.from_response(response=response,
+											formdata={'postBack':'1', 'action':'GL', 'stateSelIndex':'', 'citySelIndex':'', 'selStates':state, 'selCities':'', 'txtZipCode':'', 'selZipCodeRadius':'5' },
```
Comment (iandees) [inline locations/spiders/keyfood.py]: Please use spaces for indent instead of tabs.

File: locations/spiders/keyfood.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+				'ref': ''.join(name).replace('(', '').replace(')', '').replace(' ', '_'),
+				'street': address1,
+				'city': address2[0],
+				'state': address2[1].split(' ')[1],
+				'postcode': address2[1].split(' ')[2],
+				'opening_hours': store_hours,
+				'phone': ''.join(phone)[7:]
+			}
+
+			# print(random.randint(1,1000000))
```
Comment (iandees) [inline locations/spiders/keyfood.py]: Remove commented out debug code.

File: locations/spiders/keyfood.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+											dont_filter=True,
+											callback=self.parse_state)
+	
+	def parse_state(self,response):
+		sel = Selector(response)
+		stores = sel.css('div.StoreBox')
+
+		for store in stores:
+
+
```
Comment (iandees) [inline locations/spiders/keyfood.py]: Remove extraneous whitespace throughout file.

**Comment by bealbrown** [comment]
Howdy Ian. Made the changes you asked for. Looks like the bot timed out again. Is that an error on my end? 

Also, is there a specification for the GeojsonPointItem dict that indicates what format is acceptable for each element?

File: locations/spiders/keyfood.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        
+        for state in states:
+            yield scrapy.FormRequest.from_response(response=response,
+                                            formdata={'postBack':'1', 'action':'GL', 'stateSelIndex':'', 'citySelIndex':'', 'selStates':state, 'selCities':'', 'txtZipCode':'', 'selZipCodeRadius':'5' },
+                                            clickdata={'class': 'submitButton'},
+                                            dont_filter=True,
+                                            callback=self.parse_state)
+
+    def parse_state(self,response):
+        sel = Selector(response)
```
Comment (iandees) [inline locations/spiders/keyfood.py]: No need to use `Selector` here. Just call `response.css()`

**Comment by iandees** [comment]
The timeout is probably their server taking a long time to respond to your queries.

There's no documentation for the fields yet, but there will be. `lat` and `lon` should be `float`s, `opening_hours` should follow [OSM's opening_hours format](https://wiki.openstreetmap.org/wiki/Key:opening_hours#Examples). Otherwise they should all be strings and mostly self-explanatory. Is there a particular field you have a question about?

**Pull request merged**